File: Ch. 6/Bernulli Randomization/code/bernoulli_randomization.py
# =============================================================================
# Bernoulli Randomization
# =============================================================================
# Implements Bernoulli randomization to assign treatment status to observations
# in a dataset. Each unit is independently assigned to treatment with probability p
# and to control with probability 1-p.

from pathlib import Path
import numpy as np
import pandas as pd


# --- Setup -------------------------------------------------------------------
# Automatically resolve paths relative to this script's location.
SCRIPT_DIR = Path(__file__).resolve().parent
DATA_DIR = SCRIPT_DIR.parent / "data"
OUTPUT_DIR = SCRIPT_DIR.parent / "output"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


# --- Bernoulli randomization function ----------------------------------------
def bernoulli_randomize(data, probability, seed=None):
    """
    Assign treatment status using Bernoulli randomization.

    Each observation is independently assigned to treatment with probability p
    and to control with probability 1-p. This allows the number of treated
    units to vary randomly (unlike complete randomization which fixes it).

    Parameters
    ----------
    data : pd.DataFrame
        Input dataset to which treatment will be assigned.
    probability : float
        Probability of assignment to treatment (must be between 0 and 1).
    seed : int, optional
        Random seed for reproducibility. If None, randomization is not reproducible.

    Returns
    -------
    pd.DataFrame
        Dataset with new 'Treatment' variable (1 = treatment, 0 = control).

    Raises
    ------
    ValueError
        If probability is not between 0 and 1.
    """
    # Validate probability parameter
    if not 0 <= probability <= 1:
        raise ValueError(f"Probability must be between 0 and 1, got {probability}")

    # Create a copy to avoid modifying the original dataset
    randomized_data = data.copy()

    # Set random seed if provided
    if seed is not None:
        np.random.seed(seed)

    # Generate random draws from uniform distribution
    # Assign treatment if random draw < probability
    random_draws = np.random.uniform(0, 1, size=len(randomized_data))
    randomized_data['Treatment'] = (random_draws < probability).astype(int)

    return randomized_data


# --- Load data ---------------------------------------------------------------
# Load input dataset
# TODO: Replace 'unique_data_clean_main_synthetic.dta' with your actual dataset filename
input_file = DATA_DIR / "unique_data_clean_main_synthetic.dta"
data = pd.read_stata(input_file)


# --- Apply Bernoulli randomization -------------------------------------------
# Set treatment probability (default: 0.5 for equal allocation)
TREATMENT_PROBABILITY = 0.5

# Apply randomization with fixed seed for reproducibility
randomized_data = bernoulli_randomize(
    data=data,
    probability=TREATMENT_PROBABILITY,
    seed=None
)


# --- Print summary statistics ------------------------------------------------
n_total = len(randomized_data)
n_treatment = randomized_data['Treatment'].sum()
n_control = n_total - n_treatment
actual_proportion = n_treatment / n_total

print("\n" + "=" * 80)
print("BERNOULLI RANDOMIZATION SUMMARY")
print("=" * 80)
print(f"Treatment probability (p):        {TREATMENT_PROBABILITY:.3f}")
print(f"Total observations:                {n_total:,}")
print(f"Assigned to treatment:             {n_treatment:,}")
print(f"Assigned to control:               {n_control:,}")
print(f"Actual treatment proportion:       {actual_proportion:.3f}")
print(f"Expected number treated (n*p):     {n_total * TREATMENT_PROBABILITY:.1f}")
print("=" * 80)


# --- Save randomized dataset -------------------------------------------------
# Save to output directory
# TODO: Replace 'randomized_dataset.dta' with your desired output filename
output_file = OUTPUT_DIR / "randomized_dataset.dta"
randomized_data.to_stata(output_file, write_index=False)

print(f"\n✓ Saved to: {output_file}\n")

# =============================================================================
# END OF BERNOULLI RANDOMIZATION
# =============================================================================

File: Ch. 6/Bernulli Randomization/code/bernoulli_randomization.R
# =============================================================================
# Bernoulli Randomization
# =============================================================================
# Implements Bernoulli randomization to assign treatment status to observations
# in a dataset. Each unit is independently assigned to treatment with probability p
# and to control with probability 1-p.

# --- Setup -------------------------------------------------------------------
rm(list = ls())

# Load required libraries
library(tidyverse)
library(haven)

# Automatically set working directory to the folder containing this script.
# Supports RStudio (Source/Knit), source(), and Rscript from the command line.
if (requireNamespace("rstudioapi", quietly = TRUE) && rstudioapi::isAvailable()) {
  setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
} else if (length(sys.frames()) > 0 && !is.null(sys.frame(1)$ofile)) {
  setwd(dirname(sys.frame(1)$ofile))
} else if (!is.null(commandArgs(trailingOnly = FALSE))) {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    setwd(dirname(normalizePath(sub("^--file=", "", file_arg))))
  }
}

# Define paths relative to script location
data_dir <- file.path(dirname(getwd()), "data")
output_dir <- file.path(dirname(getwd()), "output")
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)


# --- Parameters --------------------------------------------------------------
# Set treatment probability (default: 0.5 for equal allocation)
treatment_probability <- 0.5

# Set random seed for reproducibility
# set.seed(42)


# --- Load data ---------------------------------------------------------------
# Load input dataset
# TODO: Replace 'unique_data_clean_main_synthetic.dta' with your actual dataset filename
data <- read_dta(file.path(data_dir, "unique_data_clean_main_synthetic.dta"))


# --- Apply Bernoulli randomization -------------------------------------------
# Generate random uniform draws between 0 and 1
# Assign treatment if random draw < probability
randomized_data <- data %>%
  mutate(Treatment = as.integer(runif(n()) < treatment_probability))


# --- Print summary statistics ------------------------------------------------
# Count observations by treatment status
n_total <- nrow(randomized_data)
n_treatment <- sum(randomized_data$Treatment)
n_control <- n_total - n_treatment
actual_proportion <- n_treatment / n_total
expected_treated <- n_total * treatment_probability

cat("\n", strrep("=", 80), "\n", sep = "")
cat("BERNOULLI RANDOMIZATION SUMMARY\n")
cat(strrep("=", 80), "\n", sep = "")
cat(sprintf("Treatment probability (p):        %.3f\n", treatment_probability))
cat(sprintf("Total observations:                %s\n", format(n_total, big.mark = ",")))
cat(sprintf("Assigned to treatment:             %s\n", format(n_treatment, big.mark = ",")))
cat(sprintf("Assigned to control:               %s\n", format(n_control, big.mark = ",")))
cat(sprintf("Actual treatment proportion:       %.3f\n", actual_proportion))
cat(sprintf("Expected number treated (n*p):     %.1f\n", expected_treated))
cat(strrep("=", 80), "\n", sep = "")


# --- Save randomized dataset -------------------------------------------------
# Save to output directory
# TODO: Replace 'randomized_dataset.dta' with your desired output filename
output_file <- file.path(output_dir, "randomized_dataset.dta")
write_dta(randomized_data, output_file)

cat(sprintf("\n✓ Saved to: %s\n\n", output_file))

# =============================================================================
# END OF BERNOULLI RANDOMIZATION
# =============================================================================
